{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exercise 4 - Nested Parallelism\n",
    "\n",
    "**GOAL:** The goal of this exercise is to show how to create nested tasks by calling a remote function inside of another remote function.\n",
    "\n",
    "In this exercise, you will implement the structure of a parallel hyperparameter sweep which trains a number of models in parallel. Each model will be trained using parallel gradient computations.\n",
    "\n",
    "### Concepts for this Exercise - Nested Remote Functions\n",
    "\n",
    "Remote functions can call other functions. For example, consider the following.\n",
    "\n",
    "```python\n",
    "@ray.remote\n",
    "def f():\n",
    "    return 1\n",
    "\n",
    "@ray.remote\n",
    "def g():\n",
    "    # Call f 4 times and return the resulting object IDs.\n",
    "    return [f.remote() for _ in range(4)]\n",
    "\n",
    "@ray.remote\n",
    "def h():\n",
    "    # Call f 4 times, block until those 4 tasks finish,\n",
    "    # retrieve the results, and return the values.\n",
    "    return ray.get([f.remote() for _ in range(4)])\n",
    "```\n",
    "\n",
    "Then calling `g` and `h` produces the following behavior.\n",
    "\n",
    "```python\n",
    ">>> ray.get(g.remote())\n",
    "[ObjectID(b1457ba0911ae84989aae86f89409e953dd9a80e),\n",
    " ObjectID(7c14a1d13a56d8dc01e800761a66f09201104275),\n",
    " ObjectID(99763728ffc1a2c0766a2000ebabded52514e9a6),\n",
    " ObjectID(9c2f372e1933b04b2936bb6f58161285829b9914)]\n",
    "\n",
    ">>> ray.get(h.remote())\n",
    "[1, 1, 1, 1]\n",
    "```\n",
    "\n",
    "**One limitation** is that the definition of `f` must come before the definitions of `g` and `h` because as soon as `g` is defined, it will be pickled and shipped to the workers, and so if `f` hasn't been defined yet, the definition will be incomplete."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-03-29T03:20:04.550982Z",
     "start_time": "2018-03-29T03:20:03.640299Z"
    }
   },
   "outputs": [],
   "source": [
    "from __future__ import absolute_import\n",
    "from __future__ import division\n",
    "from __future__ import print_function\n",
    "\n",
    "import numpy as np\n",
    "import ray\n",
    "import time\n",
    "\n",
    "import builtins\n",
    "\n",
    "\n",
    "@ray.remote\n",
    "def sum(xs):\n",
    "    return builtins.sum(xs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-03-29T03:20:06.170253Z",
     "start_time": "2018-03-29T03:20:05.053038Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Waiting for redis server at 127.0.0.1:58085 to respond...\n",
      "Waiting for redis server at 127.0.0.1:61938 to respond...\n",
      "Starting local scheduler with the following resources: {'CPU': 9, 'GPU': 0}.\n",
      "\n",
      "======================================================================\n",
      "View the web UI at http://localhost:8891/notebooks/ray_ui34090.ipynb?token=6fce3b4daf6ee700f16eb460c30e8e429084d55346444591\n",
      "======================================================================\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'local_scheduler_socket_names': ['/tmp/scheduler51370718'],\n",
       " 'node_ip_address': '127.0.0.1',\n",
       " 'object_store_addresses': [ObjectStoreAddress(name='/tmp/plasma_store7217893', manager_name='/tmp/plasma_manager69299679', manager_port=10221)],\n",
       " 'redis_address': '127.0.0.1:58085',\n",
       " 'webui_url': 'http://localhost:8891/notebooks/ray_ui34090.ipynb?token=6fce3b4daf6ee700f16eb460c30e8e429084d55346444591'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ray.init(num_cpus=9, redirect_output=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This example represents a hyperparameter sweep in which multiple models are trained in parallel. Each model training task also performs data parallel gradient computations.\n",
    "\n",
    "**EXERCISE:** Turn `compute_gradient` and `train_model` into remote functions so that they can be executed in parallel. Inside of `train_model`, do the calls to `compute_gradient` in parallel and fetch the results using `ray.get`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-03-29T03:24:54.585544Z",
     "start_time": "2018-03-29T03:24:54.552091Z"
    }
   },
   "outputs": [],
   "source": [
    "@ray.remote\n",
    "def compute_gradient(data):\n",
    "    time.sleep(0.03)\n",
    "    return 1\n",
    "\n",
    "\n",
    "@ray.remote\n",
    "def train_model(hyperparameters):\n",
    "\n",
    "    # EXERCISE: After you turn \"compute_gradient\" into a remote function,\n",
    "    # you will need to call it with \".remote\". The results must be retrieved\n",
    "    # with \"ray.get\" before \"sum\" is called.\n",
    "[[    sum.remote([compute_gradient.remote(i) for i in range(2)])]\n",
    "for _ in range(10)]\n",
    "    result = sum.remote(\n",
    "        [sum.remote([compute_gradient.remote(j) for j in range(2)])\n",
    "        for _ in range(10)])\n",
    "    return result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**EXERCISE:** The code below runs 3 hyperparameter experiments. Change this to run the experiments in parallel."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-03-29T03:24:57.713797Z",
     "start_time": "2018-03-29T03:24:55.651736Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Remote function \u001b[31m__main__.train_model\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/alokbeniwal/Library/Python/3.6/lib/python/site-packages/ray/worker.py\", line 790, in _process_task\n",
      "    self._store_outputs_in_objstore(return_object_ids, outputs)\n",
      "  File \"/Users/alokbeniwal/Library/Python/3.6/lib/python/site-packages/ray/worker.py\", line 717, in _store_outputs_in_objstore\n",
      "    self.put_object(objectids[i], outputs[i])\n",
      "  File \"/Users/alokbeniwal/Library/Python/3.6/lib/python/site-packages/ray/worker.py\", line 349, in put_object\n",
      "    raise Exception(\"Calling 'put' on an ObjectID is not allowed \"\n",
      "Exception: Calling 'put' on an ObjectID is not allowed (similarly, returning an ObjectID from a remote function is not allowed). If you really want to do this, you can wrap the ObjectID in a list and call 'put' on it (or return it).\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.train_model\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/alokbeniwal/Library/Python/3.6/lib/python/site-packages/ray/worker.py\", line 790, in _process_task\n",
      "    self._store_outputs_in_objstore(return_object_ids, outputs)\n",
      "  File \"/Users/alokbeniwal/Library/Python/3.6/lib/python/site-packages/ray/worker.py\", line 717, in _store_outputs_in_objstore\n",
      "    self.put_object(objectids[i], outputs[i])\n",
      "  File \"/Users/alokbeniwal/Library/Python/3.6/lib/python/site-packages/ray/worker.py\", line 349, in put_object\n",
      "    raise Exception(\"Calling 'put' on an ObjectID is not allowed \"\n",
      "Exception: Calling 'put' on an ObjectID is not allowed (similarly, returning an ObjectID from a remote function is not allowed). If you really want to do this, you can wrap the ObjectID in a list and call 'put' on it (or return it).\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.train_model\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/alokbeniwal/Library/Python/3.6/lib/python/site-packages/ray/worker.py\", line 790, in _process_task\n",
      "    self._store_outputs_in_objstore(return_object_ids, outputs)\n",
      "  File \"/Users/alokbeniwal/Library/Python/3.6/lib/python/site-packages/ray/worker.py\", line 717, in _store_outputs_in_objstore\n",
      "    self.put_object(objectids[i], outputs[i])\n",
      "  File \"/Users/alokbeniwal/Library/Python/3.6/lib/python/site-packages/ray/worker.py\", line 349, in put_object\n",
      "    raise Exception(\"Calling 'put' on an ObjectID is not allowed \"\n",
      "Exception: Calling 'put' on an ObjectID is not allowed (similarly, returning an ObjectID from a remote function is not allowed). If you really want to do this, you can wrap the ObjectID in a list and call 'put' on it (or return it).\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n"
     ]
    },
    {
     "ename": "RayGetError",
     "evalue": "Could not get objectid ObjectID(f420d9ba47fed2978384e7c5f6d38b960f2d75d4). It was created by remote function \u001b[31m__main__.train_model\u001b[39m which failed with:\n\nRemote function \u001b[31m__main__.train_model\u001b[39m failed with:\n\nTraceback (most recent call last):\n  File \"/Users/alokbeniwal/Library/Python/3.6/lib/python/site-packages/ray/worker.py\", line 790, in _process_task\n    self._store_outputs_in_objstore(return_object_ids, outputs)\n  File \"/Users/alokbeniwal/Library/Python/3.6/lib/python/site-packages/ray/worker.py\", line 717, in _store_outputs_in_objstore\n    self.put_object(objectids[i], outputs[i])\n  File \"/Users/alokbeniwal/Library/Python/3.6/lib/python/site-packages/ray/worker.py\", line 349, in put_object\n    raise Exception(\"Calling 'put' on an ObjectID is not allowed \"\nException: Calling 'put' on an ObjectID is not allowed (similarly, returning an ObjectID from a remote function is not allowed). If you really want to do this, you can wrap the ObjectID in a list and call 'put' on it (or return it).\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRayGetError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-13-1e83ae91230b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     20\u001b[0m ]\n\u001b[1;32m     21\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 22\u001b[0;31m \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mray\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtrain_model\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremote\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mh\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mh\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mhyperparameters\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     23\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0mend_time\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.6/lib/python/site-packages/ray/worker.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(object_ids, worker)\u001b[0m\n\u001b[1;32m   2243\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2244\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mRayTaskError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2245\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mRayGetError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject_ids\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2246\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2247\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRayGetError\u001b[0m: Could not get objectid ObjectID(f420d9ba47fed2978384e7c5f6d38b960f2d75d4). It was created by remote function \u001b[31m__main__.train_model\u001b[39m which failed with:\n\nRemote function \u001b[31m__main__.train_model\u001b[39m failed with:\n\nTraceback (most recent call last):\n  File \"/Users/alokbeniwal/Library/Python/3.6/lib/python/site-packages/ray/worker.py\", line 790, in _process_task\n    self._store_outputs_in_objstore(return_object_ids, outputs)\n  File \"/Users/alokbeniwal/Library/Python/3.6/lib/python/site-packages/ray/worker.py\", line 717, in _store_outputs_in_objstore\n    self.put_object(objectids[i], outputs[i])\n  File \"/Users/alokbeniwal/Library/Python/3.6/lib/python/site-packages/ray/worker.py\", line 349, in put_object\n    raise Exception(\"Calling 'put' on an ObjectID is not allowed \"\nException: Calling 'put' on an ObjectID is not allowed (similarly, returning an ObjectID from a remote function is not allowed). If you really want to do this, you can wrap the ObjectID in a list and call 'put' on it (or return it).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n",
      "Remote function \u001b[31m__main__.sum\u001b[39m failed with:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-11-0ba2ad0ebe88>\", line 8, in sum\n",
      "TypeError: unsupported operand type(s) for +: 'int' and 'common.ObjectID'\n",
      "\n",
      "\n",
      "  You can inspect errors by running\n",
      "\n",
      "      ray.error_info()\n",
      "\n",
      "  If this driver is hanging, start a new one with\n",
      "\n",
      "      ray.init(redis_address=\"127.0.0.1:58085\")\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "# Sleep a little to improve the accuracy of the timing measurements below.\n",
    "time.sleep(2.0)\n",
    "start_time = time.time()\n",
    "\n",
    "# Run some hyperparaameter experiments.\n",
    "\n",
    "hyperparameters = [\n",
    "    {\n",
    "        'learning_rate': 1e-1,\n",
    "        'batch_size': 100,\n",
    "    },\n",
    "    {\n",
    "        'learning_rate': 1e-2,\n",
    "        'batch_size': 100,\n",
    "    },\n",
    "    {\n",
    "        'learning_rate': 1e-3,\n",
    "        'batch_size': 100,\n",
    "    },\n",
    "]\n",
    "\n",
    "results = ray.get([train_model.remote(h) for h in hyperparameters])\n",
    "\n",
    "end_time = time.time()\n",
    "duration = end_time - start_time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**VERIFY:** Run some checks to verify that the changes you made to the code were correct. Some of the checks should fail when you initially run the cells. After completing the exercises, the checks should pass."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-03-28T05:28:56.086271Z",
     "start_time": "2018-03-28T05:28:56.062787Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Success! The example took 0.36385107040405273 seconds.\n"
     ]
    }
   ],
   "source": [
    "assert results == [20, 20, 20]\n",
    "assert duration < 0.5, ('The experiments ran in {} seconds. This is too '\n",
    "                         'slow.'.format(duration))\n",
    "assert duration > 0.3, ('The experiments ran in {} seconds. This is too '\n",
    "                        'fast.'.format(duration))\n",
    "\n",
    "print('Success! The example took {} seconds.'.format(duration))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**EXERCISE:** Use the UI to view the task timeline and to verify that the pattern makes sense."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-03-28T05:28:59.080371Z",
     "start_time": "2018-03-28T05:28:58.932476Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "To view fullscreen, open chrome://tracing in Google Chrome and load `/var/folders/52/pd9v2v217dqdkj574btgbn4h0000gp/T/tmpbgkqyaho.json`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"900\"\n",
       "            height=\"800\"\n",
       "            src=\"tmp25tx7o97.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x1175e3470>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import ray.experimental.ui as ui\n",
    "ui.task_timeline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
